File: ruby/beer-song/beer_song.rb
class BeerSong
  def verse(number)
    phraseN(number)
  end

  def verses(start, stop)
    # Ranges won't work descending.
    start.downto(stop).collect { |n| verse(n) }.join("\n")
  end

  def lyrics
    verses(99, 0)
  end

  private

  def phraseN(n)
    return phrase0 if n == 0
    return phrase1 if n == 1
    "#{n} bottles of beer on the wall, #{n} bottles of beer.\nTake one down and pass it around, #{n - 1} bottle#{n - 1 == 1 ? '' : 's'} of beer on the wall.\n"
  end

  def phrase0
    "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"
  end

  def phrase1
    "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n"
  end
end

module BookKeeping
  VERSION = 2
end

File: typescript/beer-song/beer-song.ts
const phrase0 = (): string => { return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n" };
const phrase1 = (): string => { return "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n" };
const phraseN = (n: number): string => { return `${n} bottles of beer on the wall, ${n} bottles of beer.\nTake one down and pass it around, ${n - 1} bottle${n - 1 == 1 ? '' : 's'} of beer on the wall.\n` };

export default class Beer {
  static verse(number: number): string {
    switch (number) {
      case 0: return phrase0();
      case 1: return phrase1();
      default: return phraseN(number);
    }
  }

  static sing(start = 99, end = 0): string {
    let out = [];
    for (var x = start; x >= end; x--) {
      out.push(this.verse(x));
    }
    return out.join("\n");
  };
}